File: frontend/playback.js
// playback.js — PlaybackController
export class PlaybackController {
  constructor(project, onTick) {
    this.project = project;
    this._onTick = onTick;
    this.playing = false;
    this.playhead = 0.0;
    this.loop = false; // toggle via .setLoop(bool)
    this._raf = null;
    this._lastTs = null;
  }

  setProject(p) {
    this.project = p;
    this.playhead = Math.min(this.playhead, p.duration);
  }

  setLoop(on) { this.loop = on; }
  toggle() { this.playing ? this.pause() : this._play(); }

  pause() {
    this.playing = false;
    if (this._raf) { cancelAnimationFrame(this._raf); this._raf = null; }
    this._lastTs = null;
  }

  seek(t) {
    this.playhead = Math.max(0, Math.min(t, this.project.duration));
    this._onTick(this.playhead, true); // true = explicit seek, not a playback tick
  }

  seekRelative(dt) { this.seek(this.playhead + dt); }

  stepFrame(dir = 1) {
    const fps = this.project.fps ?? 30;
    this.seekRelative(dir / fps);
  }

  _play() {
    if (this.playing) return;
    if (this.playhead >= this.project.duration) this.playhead = 0.0;
    this.playing = true;
    this._lastTs = null;
    this._raf = requestAnimationFrame(ts => this._tick(ts));
  }

  _tick(ts) {
    if (!this.playing) return;

    if (this._lastTs !== null) {
      const dt = (ts - this._lastTs) / 1000.0;
      this.playhead = this.playhead + dt;

      if (this.playhead >= this.project.duration) {
        if (this.loop) {
          this.playhead = this.playhead % this.project.duration;
        } else {
          this.playhead = this.project.duration;
          this._onTick(this.playhead);
          this.pause();
          return;
        }
      }
      this._onTick(this.playhead);
    }

    this._lastTs = ts;
    this._raf = requestAnimationFrame(ts2 => this._tick(ts2));
  }
}

File: backend/playback.py
import time


class PlaybackController:

    def __init__(self, project):
        self.project = project
        self.playhead = 0.0
        self.playing = False

        self._last_tick = None

    def play(self):
        self.playing = True
        self._last_tick = time.time()

    def pause(self):
        self.playing = False

    def toggle(self):
        if self.playing:
            self.pause()
        else:
            self.play()

    def seek(self, t):
        self.playhead = max(
            0.0,
            min(t, self.project.duration)
        )

    def tick(self):
        if not self.playing:
            return self.playhead

        now = time.time()
        dt = now - self._last_tick
        self._last_tick = now

        self.playhead += dt

        if self.playhead >= self.project.duration:
            self.playhead = 0.0
            self.pause()

        return self.playhead